pkg/expression/constant.go

```diff
@@ -135,15 +135,20 @@ type ParamMarker struct {
 }
 
 // GetUserVar returns the corresponding user variable presented in the `EXECUTE` statement or `COM_EXECUTE` command.
-func (d *ParamMarker) GetUserVar() types.Datum {
+func (d *ParamMarker) GetUserVar(ctx EvalContext) types.Datum {
+	return ctx.GetParamValue(d.order)
```

Comment:
Why don't we just consider `EvalContext` as a field of `ParamMarker` like `d.ctx`? Then we can remove the call like `ctx.GetExprCtx().GetEvalCtx()`.

Reply:
Because the goal of this PR is to finally remove `d.ctx`. To remove the `d.ctx`, it's required to add a `ctx EvalContext` argument to each method. It'll be a big change, so this PR only modifies `GetType` method.

This PR didn't remove the `d.ctx` because the `String()` method is not modified, and is still using the `d.ctx`. It'll finally be removed in https://github.com/pingcap/tidb/pull/53534 (which has two more commits based on this change).